crates/wasmtime/src/runtime/vm/libcalls.rs

```diff
@@ -1137,6 +1137,40 @@ unsafe fn check_malloc(
     Ok(0)
 }
 
+// Hook for validating calloc using wmemcheck_state.
+#[cfg(feature = "wmemcheck")]
+unsafe fn check_calloc(store: &mut dyn VMStore, instance: &mut Instance, addr: u32, count: u32, size: u32) -> Result<u32> {
+    check_malloc(store, instance, addr, count * size)
+}
+
+// Hook for validating realloc using wmemcheck_state.
+#[cfg(feature = "wmemcheck")]
+unsafe fn check_realloc(store: &mut dyn VMStore, instance: &mut Instance, end_addr: u32, start_addr: u32, len: u32) -> Result<u32> {
+    check_free(store, instance, start_addr)?;
+    check_malloc(store, instance, end_addr, len)
+}
+
+// Hook for validating malloc_usable_size using wmemcheck_state.
+#[cfg(feature = "wmemcheck")]
+unsafe fn check_malloc_usable_size(store: &mut dyn VMStore, instance: &mut Instance, len: u32, addr: u32) -> Result<u32> {
+    check_free(store, instance, addr)?;
+    check_malloc(store, instance, addr, len)
+}
+
+
+// Hook for validating posix_memalign using wmemcheck_state.
+#[cfg(feature = "wmemcheck")]
+unsafe fn check_posix_memalign(store: &mut dyn VMStore, instance: &mut Instance, outptr: u32, size: u32) -> Result<u32> {
+    for (_, entry) in instance.exports() {
+        if let wasmtime_environ::EntityIndex::Memory(mem_index) = entry {
+            let mem = instance.get_memory(*mem_index);
+            let out = *(mem.base.offset(outptr as isize) as *mut u32);
```

Comment:
We definitely need to do a bounds-check here -- could you use the safe APIs for the memory to access the out-pointer?

This is also little-endian-specific code; we'll want to get a `[u8; 4]` and do an explicit conversion with `u32::from_le_bytes`. (Wasmtime supports one big-endian architecture, s390x!)